### PR TITLE
consumer: flexible transaction boundaries

### PR DESCRIPTION
**Pull request by jgraettinger**
Update `message.Sequencer` to enable fine-grain tracking within a run of acknowledged messages. The Sequencer is now able to take checkpoints that reflect partial processing of a single run, by incrementally tightening the minimum Clock and offset of the ProducerState as messages are processed.

Update `consumer` transaction processing to hoist the former tight loop of processing an acknowledged message run, to the top-level `txnStep`. This makes processing of runs incremental, and subject to interruption due to transaction timeouts or a new error return-able by clients, electing to defer processing of a message to a following transaction.

Tested using Gazette's soak tests, and also within Flow.

Issue #300

---
This change is [Reviewable](https://reviewable.io/reviews/gazette/core/303)

### CODE REVIEW

**Comment by jgraettinger** [comment]
@psFried  FYI

Some more unit testing and a lot of integration testing I want to do, plus reworking commit structure, but I believe the implementation is all here.

File: consumer/transaction.go
Comment (psFried) [inline consumer/transaction.go:240]: Just checking, do we need to call `Step` in a loop here, or does one call take care of all the messages implied by `QueueAckCommitRing`?

File: consumer/transaction.go
Comment (psFried) [inline consumer/transaction.go:215]: There's a lot of these `QueueOutcome` constants, and it's kind of tough as a reader to see that we're handling all of these correctly in terms of the contract with the `Sequencer`. One suggestion would be to define functions like `func (o QueueOutcome) MustStep() bool` make that more clear.

File: consumer/transaction.go
Comment (psFried) [inline consumer/transaction.go:305]: Seems like maybe we should wait to update these stats until after we know the message isn't getting deferred until the next transaction.

**Review by psFried — COMMENTED** [review]
Leaving some thoughts, but overall I feel like this all makes sense and seems right. This is serving as a marker of where I left off in review, since I think you're still adding some commits.

File: consumer/protocol/checkpoint_extensions.go
Comment (jgraettinger) [reply consumer/protocol/checkpoint_extensions.go:22]: You're right, backing out.

File: consumer/protocol/checkpoint_extensions.go
Comment (jgraettinger) [reply consumer/protocol/checkpoint_extensions.go:36]: `source` isn't a pointer but a value copy, so this is actually updating the map. Commented.

I changed Source to be non-nullable, and Go doesn't let you mutate an inner field of a map element (you can only set the whole value).

File: consumer/transaction.go
Comment (jgraettinger) [reply consumer/transaction.go:305]: Done.

File: consumer/transaction.go
Comment (jgraettinger) [reply consumer/transaction.go:215]: You know, I had done that but ended up backing it out, because you're always allowed to call `Step` and the simpler cases just do that. Otherwise the only QueueOutcome you need to pay attention to is `QueueAckCommitReplay`, because its on you to start a replay read. `transaction.go` cares about more of these, because it's doing fancier stuff with respect to reporting progress. So I backed it out because there isn't actually a common thread on how it'd be used.

File: consumer/transaction.go
Comment (jgraettinger) [reply consumer/transaction.go:240]: No, we definitely want to step _just one_ message because consumption must be driven by the top-level `txnStep`, which is checking for a timeout or for the consumer app to defer the message or lots of other things.

Basically everything got inverted, and the tight loop that used to process committed messages instead now makes progress by single polled calls to `txnStep`, each of which handles a single dequeued message. The trade-off is that we can precisely end the transaction whenever we want (timeout, or because the app deferred the message).

File: consumer/transaction.go
Comment (jgraettinger) [reply consumer/transaction.go:131]: Done

**Comment by jgraettinger** [comment]
I've addressed feedback and re-worked commit messages. I left all but the first two commits (which were merged up WIP) alone.

**Review by psFried — APPROVED** [review]
I think this is looking good. I feel like it would be good to get another set of eyes on this before merging, though, since it's a pretty tricky change, and I'm not 100% clear on to what extent these changes affect portions of the API that people are using today.

File: consumer/shard.go
Comment (saterus) [inline consumer/shard.go:244]: When a consumer signals they would like to close the transaction via `ErrDeferToNextTransaction`, we nil out the `txn.readCh`.
1. Do we need to explicitly close this channel?
2. Is this the only mechanism needed to resume? Or is there more to the contract around using `ErrDeferToNextTransaction`?

On the Flow side, will this cause any `Task`/`Application` to restart itself when the channel is closed and we subsequently call `StartReadingMessages` again? I'm looking at `Captures` since it's a `MessageProducer`, but this could matter for non-`MessageProducer`s too.

Specifically, `Capture.serverDriverTransactions`, which will attempt to send data to this channel that we're setting to `nil`. I'm not sure how (or how far) this propagates back to the Capture driver, but it seems like we either need to maintain some continuity or deliberately restart things.

**Review by saterus — COMMENTED** [review]
I can't say I'm an expert at any of this yet, but this exercise definitely helped connect a lot of dots between abstract concepts and the code that makes it real. I have a few lingering questions that you might be able to help clear up.

File: message/sequencer.go
Comment (jgraettinger) [reply message/sequencer.go:49]: The `Dequeued` message is passed to the application's `ConsumeMessage` callback, and that can do whatever it likes with the message. We can't _assume_ that its UUID continues to be the UUID we read. And in fact it won't be, if the application immediately published the message to another journal (as that sets a new UUID value on the decoded instance).

File: consumer/transaction.go
Comment (jgraettinger) [reply consumer/transaction.go:133]: Ah! This is a neat detail of the Go spec and a fun technique.

A `var foo chan Bar` is always a _reference_ to an allocated channel structure, or nil. You can swap out the backing channel it points to, or set it to `nil` to nil out the reference _without touching the allocated channel_, which may have other references.

Selecting into or out of a nil channel blocks forever. It doesn't crash. That means that nil-ing out `foo` is a good way to turn it off. So, here we're not messing with the underlying message channel, we're just causing it to not select in this context.

File: consumer/transaction.go
Comment (jgraettinger) [reply consumer/transaction.go:161]: That'd be fine, but I made the other branch explicit for the sake of the comment.

Given an explicit branch like this, the most idiomatic version would be to [indent the error flow](https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow):

```
if err := txnConsume(s, txn); err != nil {
    return false, err
}
return false, nil // We consumed one message.
```

I don't have a good rationale for not doing it like that ⬆️

File: consumer/shard.go
Comment (jgraettinger) [reply consumer/shard.go:244]: 1) No, we don't want to close the channel. Only senders should ever close a channel in Go, and we're a receiver. In fact, `readCh` isn't a type `chan Message` but rather a type `<-chan Message`, meaning it can only receive and cannot be closed. This particular channel gets closed by the implementation of `StartReadingMessages` that pumps into it, when that function decides it's done for whatever reason and isn't going to send anymore.

2) The effect of this is to not read any more messages _in this transaction_. The `readCh` variable on the *transaction is a separate reference from the `readCh` passed to `runTransactions`. Each `transaction` starts with the latter, and sets it to nil when it's not going to read anymore. `txnInit` initializes the next `transaction` with a non-nil `readCh` and so on.

**Comment by jgraettinger** [comment]
We've been running this in production a few days and it's looking solid. Going ahead and merging.

File: message/sequencer.go
Comment (saterus) [reply message/sequencer.go:49]: It looks like [we're passing the `Envelope` to `ConsumeMessage` by value, rather than by reference](https://github.com/gazette/core/blob/75eedeeeb01bd1a85836f91760dec640d0d4ff5c/consumer/transaction.go#L305-L305), right? The `Application` shouldn't be able to mutate the message that's sitting in the `Sequencer`.

Even if we say that it could potentially mutate it and we need to defend against that here, phrasing it as "change the Application" seems a bit misleading to me. If you're republishing the same message to a different journal, it's just a different message now. This seems reflected in the reality of what's happening: the UUID of the message is changing. Is this the right way to think about this?

File: consumer/transaction.go
Comment (saterus) [reply consumer/transaction.go:133]: Ah, I see now. I didn't internalize that channels were always passed by reference. It's inconsistent since the syntax is hiding the pointer type with `var foo chan Bar` rather than `var foo *(chan Bar)` (or some syntax like that).

I knew about non-blocking reads from channels with `select`, but I don't think I put together that inside a `select` the `<-nil` would block rather than crash. Outside of `select` the `<-nil` crashes.

File: consumer/shard.go
Comment (saterus) [reply consumer/shard.go:244]: > The effect of this is to not read any more messages in this transaction. The readCh variable on the *transaction is a separate reference from the readCh passed to runTransactions. Each transaction starts with the latter, and sets it to nil when it's not going to read anymore. txnInit initializes the next transaction with a non-nil readCh and so on.

Gotcha! This is what I was missing here. The channel continues to fill with new messages, even as the transaction goes through the motions of committing itself. :100: 

> Only senders should ever close a channel in Go

How does a receiver signal that it's done? Would you open yet another channel to signal that we're hanging up?

File: message/sequencer.go
Comment (jgraettinger) [reply message/sequencer.go:49]: The `Envelope` is passed by-value, including its contained `Message` field which is an interface (always a reference). We're essentially passing a reference by-value, but it's still a reference.

File: consumer/shard.go
Comment (jgraettinger) [reply consumer/shard.go:244]: | How does a receiver signal that it's done?

This is really where `Context` comes in. You cancel a context that the sender is also selecting over.

Naively, just having the receiver go away is a good way to get leaked goroutnines trying to send into a channel that will never again unblock. So you always want another avenue by which the sender discovers that it's work isn't needed anymore. Check out the StartReadingMessages pump in go/shuffle/reader.go in the Flow codebase; it's send is a select over sending a message OR receiving <-ctx.Done().

Also check out `servePrimary` in shard.go. When we want to do a clean restart of the message loop, that's done by having the _sender_ close the channel, and when we've read through its contents and the close, then we start up a new message loop.

File: consumer/transaction.go
Comment (jgraettinger) [reply consumer/transaction.go:133]: Yep. `map` is also by-reference FYI.

**Pull request merged**
